### PR TITLE
Add documentation on custom indexes

### PR DESCRIPTION
**Pull request by benbovy**
This PR documents the API of the `Index` base class and adds a guide for creating custom indexes (reworked from https://hackmd.io/Zxw_zCa7Rbynx_iJu6Y3LA). Hopefully it will help anyone experimenting with this feature.

@pydata/xarray your feedback would be very much appreciated! I've been into this for quite some time, so there may be things that seem obvious to me but that you can still find very confusing or non-intuitive. It would then deserve some extra or better explanation.

More specifically, I'm open to any suggestion on how to better illustrate this with clear and succinct examples.

There are other parts of the documentation that still need to be updated regarding the indexes refactor (e.g., "dimension" coordinates, `xindexes` property, set/drop indexes, etc.). But I suggest to do that in separate PRs and focus here on creating custom indexes.

### CODE REVIEW

File: xarray/core/indexes.py [hunk 1]
```diff
@@ -45,72 +88,315 @@ def concat(
...
+
+        Returns
+        -------
+        sel_results : :py:class:`IndexSelResult`
+            An index query result object that contains dimension positional indexers.
+            It may also contain new indexes, coordinate variables, etc.
+        """
         raise NotImplementedError(f"{self!r} doesn't support label-based selection")
 
     def join(self: T_Index, other: T_Index, how: str = "inner") -> T_Index:
```
Comment (Illviljan) [inline xarray/core/indexes.py:262]:
```suggestion
    def join(self: T_Index, other: T_Index, how: JoinOptions = "inner") -> T_Index:
```
You can use Literals from:
https://github.com/pydata/xarray/blob/19b9e04e9a7f0118f35aa2b8946df0ef7cb3375c/xarray/core/types.py#L64

I'm not so sure about adding typing in the docstring anymore, there's a real risk that they will forgotten about and won't match the typing. 
Dask has a nice guideline with some discussion about it here:
https://github.com/dask/community/issues/255

Comment (benbovy) [reply xarray/core/indexes.py:262]: > I'm not so sure about adding typing in the docstring anymore, there's a real risk that they will forgotten about and won't match the typing.

Yes let's remove it. The goal of the Index docstrings is to help those who want to implement a custom index and I assume they are already familiar with Xarray.

**Comment by dcherian** [comment]
> an inefficient "numpy" index with basic lookup

yes! I used this recently to describe what an index does. I think most people are familiar with the argmin way

File: doc/internals/how-to-create-custom-index.rst [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+This example is also not very useful unless we add some extra functionality on
+top of the two encapsulated ``PandasIndex`` objects, such as a coordinate
+reference system.
+
+How to use a custom index
+-------------------------
+
+You can use ``Dataset.set_xindex()`` or ``DataArray.set_xindex()`` to assign a
+custom index to a Dataset or DataArray, e.g., using the ``RasterIndex`` above:
```
Comment (dopplershift) [inline doc/internals/how-to-create-custom-index.rst:213]: As far as I can find these APIs haven't actually been added?

Comment (benbovy) [reply doc/internals/how-to-create-custom-index.rst:213]: I has been added in #6971, which should probably be merged before this one. Once it is done, I'll add here proper references to the API doc.

File: xarray/core/indexes.py [hunk 3]
```diff
@@ -45,72 +88,315 @@ def concat(
...
 
     @classmethod
-    def stack(cls, variables: Mapping[Any, Variable], dim: Hashable) -> Index:
+    def stack(
+        cls: type[T_Index], variables: Mapping[Any, Variable], dim: Hashable
+    ) -> T_Index:
+        """Create a new index by stacking coordinate variables into a single new
+        dimension.
+
+        Implementation is optional but required in order to support stack.
```
Comment (dcherian) [inline xarray/core/indexes.py:124]:
```suggestion
        Implementation is optional but required in order to support ``stack``.
```

File: xarray/core/indexes.py [hunk 4]
```diff
@@ -45,72 +88,315 @@ def concat(
...
+        """Unstack a (multi-)index into multiple (single) indexes.
+
+        Implementation is optional but required in order to support unstacking
+        the coordinates from which this index has been built.
+
+        Returns
+        -------
+        indexes : tuple
+            A 2-length tuple where the 1st item is a dictionary of unstacked
+            Index objects and the 2nd item is a :py:class`pandas.MultiIndex`
```
Comment (dcherian) [inline xarray/core/indexes.py:154]:
```suggestion
            Index objects and the 2nd item is a :py:class:`pandas.MultiIndex`
```

File: xarray/core/indexes.py [hunk 5]
```diff
@@ -45,72 +88,315 @@ def concat(
...
+            A new Index object.
+        """
         raise NotImplementedError(
             f"{self!r} doesn't support alignment with inner/outer join method"
         )
 
     def reindex_like(self: T_Index, other: T_Index) -> dict[Hashable, Any]:
+        """Query the index with another index of the same type.
+
+        Implementation is optional but required in order to support alignment.
```
Comment (dcherian) [inline xarray/core/indexes.py:287]: Would be good to be specific about types of alignment here. Do we also need to mention reindexing?

File: xarray/core/indexes.py [hunk 6]
```diff
@@ -45,72 +88,315 @@ def concat(
...
+    def __deepcopy__(self: T_Index, memo=None) -> T_Index:
         # memo does nothing but is required for compatibility with
         # copy.deepcopy
         return self.copy(deep=True)
 
-    def copy(self, deep: bool = True) -> Index:
+    def copy(self: T_Index, deep: bool = True) -> T_Index:
+        """Return a (deep) copy of this index.
+
+        Implementation is subclasses of Index is optional. The base class
```
Comment (dcherian) [inline xarray/core/indexes.py:386]:
```suggestion
        Implementation in subclasses of Index is optional. The base class
```

File: xarray/core/indexes.py [hunk 7]
```diff
@@ -45,72 +88,315 @@ def concat(
...
-        self, name_dict: Mapping[Any, Hashable], dims_dict: Mapping[Any, Hashable]
-    ) -> Index:
+        self: T_Index,
+        name_dict: Mapping[Any, Hashable],
+        dims_dict: Mapping[Any, Hashable],
+    ) -> T_Index:
+        """Maybe update the index with new coordinate and dimension names.
+
+        This method should be re-implemented in subclasses of Index if it has
+        attributes that depends on coordinate or dimension names.
```
Comment (dcherian) [inline xarray/core/indexes.py:352]:
```suggestion
        attributes that depend on coordinate or dimension names.
```

File: xarray/core/indexes.py [hunk 8]
```diff
@@ -45,72 +88,315 @@ def concat(
...
+
+        By default (if not re-implemented), it returns the index itself.
+
+        Warning: the input names are not filtered for this index, they may
```
Comment (dcherian) [inline xarray/core/indexes.py:356]:
```suggestion
        Warning: the input names are not filtered for this method, they may
```

I found this warning confusing. So for rename, we call `rename` on all indexes passing the dict ? And indexes are expected to return `self` if none of the names are relevant

**Review by dcherian — COMMENTED** [review]
Nice work! 

I have just minor suggestions

**Comment by TomNicholas** [comment]
> **pydata/xarray** your feedback would be very much appreciated! I've been into this for quite some time, so there may be things that seem obvious to me but that you can still find very confusing or non-intuitive. It would then deserve some extra or better explanation.

I find the way in which the Index objects are involved in a method call (e.g. `sel`) pretty opaque still. (Bear in mind I don't think I've ever contributed a PR to xarray that touched `indexes.py` or `indexing.py` :sweat_smile:)

- When is my `CustomIndex` object consulted? Is it potentially for all basic operations (concat, join, align, indexing, etc?) If I passed some kind of `NotImplementedIndex` which _only_ defined `.from_variables`, what functionality would be left in xarray?
- Why does there not need to be an index (in `.indexes`) if I do indexing with e.g. `.isel` but have no coordinates?
- If I index along two dimensions simultaneously (`.isel(x=1, y=2)`) does that correspond to two separate Index consultations?
- How is an `xarray.PandasIndex` different from a `pd.Index`?
- When I load the "air" tutorial data and it shows a `Float64Index` and `DateTime64Index`, where did they come from?

Perhaps these questions are too specific to xarray's internals but I do think there should be some kind of mental model given as to the role the Index objects play. (This could be a white lie, similar to how our page on data structures says that `Dataset` objects contain `DataArray` objects when actually they technically don't.)

I also think we need multiple simple examples. How about
- `HelloWorldIndex`, that just prints `"I'm calling HelloWorldIndex.isel !"` etc.
- `PeriodicBoundaryIndex` (a partial/simpler implementation of #7031),
- A simple functionally-derived index, that consults a dynamically-called exponential function or something.
- Some example of a custom multi-index, perhaps some kind of 2D lat-lon thing? Or how about something that represents 2D image distortion, like using it creates a fisheye effect?

I find it helps to think of documentation using [this 4-part system](https://documentation.divio.com/). This PR should cover "Explanation" pretty well, but we should still aim for other content to better cover "Tutorial", "How-to Guides", and "Reference". "Tutorial" could be like a notebook walking through creating a simple index (e.g. `PeriodicIndex`) from scratch, explaining and fixing errors as they arise (like @dcherian did for `apply_ufunc`). "How-to Guides" might be specific to other more advanced custom index examples. I guess "Reference" here is just having really clear docstrings on the possible methods of `Index` somewhere (we can't really do that with an ABC though can we?).

That all said, this is already a great start!

**Comment by benbovy** [comment]
Thanks @dcherian and @TomNicholas for your feeback!

@dcherian I will reply to your inline comments when I'll integrate your suggestions in this PR.

@TomNicholas I answer to your comments below.

> Bear in mind I don't think I've ever contributed a PR to xarray that touched indexes.py or indexing.py

That's exactly why your feedback is valuable!

> When is my CustomIndex object consulted? Is it potentially for all basic operations (concat, join, align, indexing, etc?)

I agree this could be detailed more in the Index API docstrings in a consistent way. For some methods like `equals`, `join` and `reindex_like` it could be called in *a lot* of places, basically everything that relies on object alignment.

> Why does there not need to be an index (in .indexes) if I do indexing with e.g. .isel but have no coordinates?

We should clarify that the aim of `Index` objects is to make more efficient all the operations made in the (discrete or continuous) space defined by the coordinate labels. That space is distinct from the discrete space defined by array element locations. All operations made in the latter space don't require any index.

Some Index API like `Index.isel` suggest otherwise, but those methods are rather for convenience, i.e., avoid users having to rebuild an index from scratch when it could be easily built from the existing one.

> How is an xarray.PandasIndex different from a pd.Index?

I've tried to explain it in the "Index base class" section and the sections below, but maybe it should be emphasized more?

> When I load the "air" tutorial data and it shows a Float64Index and DateTime64Index, where did they come from?

I guess you mean it is shown through `ds.indexes`?

`ds.xindexes` (vs. `ds.indexes`) still needs to be added in the docs (in a later PR?), which hopefully will address your concern here.

> Finally it's not great that to explain Index objects we have to assume the user knows what xarray.Variable is, but Variable is still not really public API, and certainly isn't documented as comprehensively as DataArray and Dataset are.

I agree, although `Variable` is already documented in the "Xarray internals" section. Maybe `Index` also deserves its own entry there, where we could explain what indexes are, how they are different from variables (coordinates), how they are used or accessed in Xarray, etc.

Overall, I think that the whole "Xarray Internals" section could be streamlined beyond a bunch of loosely-coupled document pages.

> I also think we need multiple simple examples.

I agree that we need more examples, but I also think that too much examples may tend to make things more confused.

One thing that I like very much in https://fastapi.tiangolo.com/ is how a small example is picked for each tutorial and then is shown by highlighting the relevant code for every subsection. Is it possible to do that with Sphinx / RST?

It's hard to show all features through one succinct example, though. Like @dcherian says in https://github.com/pydata/xarray/pull/6975#discussion_r967495773, we could invite people to look into the `PandasIndex` and `PandasMultiIndex` code for more details. My hope is that there will be more real examples (multi-coordinate, multi-dimensions) available in the future.

**Comment by TomNicholas** [comment]
> We should clarify that the aim of Index objects is to make more efficient all the operations made in the (discrete or continuous) space defined by the coordinate labels. That space is distinct from the discrete space defined by array element locations.

I think this should be one of the first things said. It defines what all the following discussion of Indexes does and does not affect.

> I've tried to explain it in the "Index base class" section and the sections below, but maybe it should be emphasized more?

Yeah I think you do actually have that one covered, I just included it as another example of a naive question that everyone will have that is worth heading off very explicitly.

> When I load the "air" tutorial data and it shows a Float64Index and DateTime64Index, where did they come from?
> 
> I guess you mean it is shown through `ds.indexes`?
>
> ds.xindexes (vs. ds.indexes) still needs to be added in the docs (in a later PR?), which hopefully will address your concern here.

I meant like when did these indexes get automatically built? (Presumably on coordinate assignment)

> Maybe Index also deserves its own entry there, where we could explain what indexes are, how they are different from variables (coordinates), how they are used or accessed in Xarray, etc.

1000% yes we need a page that explains what `Index` objects are, what they do, and how they work, and how they are handled automatically by default. This is pre-requisite knowledge (which apparently I don't have :sweat_smile: ) before trying to build your own custom index.

> Overall, I think that the whole "Xarray Internals" section could be streamlined beyond a bunch of loosely-coupled document pages.

Probably, but having a loosely coupled page for each aspect of the internals would be a good initial aim.

> I agree that we need more examples, but I also think that too much examples may tend to make things more confused.

That's why I like the "Explanation" vs "How-to" vs "Tutorials" distinction: use minimal code in the "Explanation" section (this PR) but put multiple more complex examples under "How to create a functionally-derived index", "how-to create a lazy index" etc.

File: xarray/core/indexes.py [hunk 9]
```diff
@@ -25,17 +25,60 @@
...
+    When a method is not (re)implemented, depending on the case the
+    corresponding operation on a :py:meth:`Dataset` or :py:meth:`DataArray`
+    either will raise a ``NotImplementedError`` or will simply drop/pass/copy
+    the index from/to the result.
+
+    """
 
     @classmethod
-    def from_variables(cls, variables: Mapping[Any, Variable]) -> Index:
+    def from_variables(
```
Comment (TomNicholas) [inline xarray/core/indexes.py:60]: Tangential but couldn't this be decorated to make it an abstract (class) method? Then an error would be raised if subclasses don't implement it. i.e. this:

```python
from abc import ABC, abstractmethod

class Index(ABC):
     @classmethod
     @abstractmethod
     def from_variables(cls, vars):
         ...

     def sel(self, indexers):
         raise NotImplementedError()
 
```

**Comment by jsignell** [comment]
Just a note that there are still some unmerged suggestions that could easily be incorporated before merge.

**Comment by TomNicholas** [comment]
@benbovy do you mind if we merge this and leave the comments to address in a follow-up PR? At SciPy we met some advanced developers from other fields (core devs of napari and astropy) who are very interested in creating functional indexes for their use cases, so it would be nice to merge this.

**Comment by benbovy** [comment]
@TomNicholas I don't mind at all! Let's merge this even though it is not perfect. I'll improve it and address the comments made here in a follow-up PR. I'll also add `Index` and `Coordinates` (once #7368 is merged) sections in the internals page added in #7991.

**Comment by benbovy** [comment]
> advanced developers from other fields (core devs of napari and astropy) who are very interested in creating functional indexes for their use cases

There is an example of a functional index in this discussion (last item): https://github.com/pydata/xarray/discussions/7041#discussion-4394875

File: doc/internals/how-to-create-custom-index.rst [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+Every index must at least implement the :py:meth:`Index.from_variables` class
+method, which is used by Xarray to build a new index instance from one or more
+existing coordinates in a Dataset or DataArray.
+
+Since any collection of coordinates can be passed to that method (i.e., the
+number, order and dimensions of the coordinates are all arbitrary), it is the
+responsibility of the index to check the consistency and validity of those input
+coordinates.
+
+For example, ``PandasIndex`` accepts only one coordinate and
```
Comment (TomNicholas) [inline doc/internals/how-to-create-custom-index.rst:54]:
```suggestion
For example, :py:class:`~xarray.core.indexes.PandasIndex` accepts only one coordinate and
```

File: doc/internals/how-to-create-custom-index.rst [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+``PandasMultiIndex`` accepts one or more 1-dimensional coordinates that must all
```
Comment (TomNicholas) [inline doc/internals/how-to-create-custom-index.rst:55]:
```suggestion
For example, :py:class:`~xarray.core.indexes.PandasIndex` accepts only one coordinate and
:py:class:`~xarray.core.indexes.PandasMultiIndex` accepts one or more 1-dimensional coordinates that must all
```

**Pull request merged**